### ISSUE TITLE
str.starts_with doesn't filter small prefixes in longer strings correctly

### ISSUE THREAD

**Issue by harm-matthias-harms** [pola-rs/polars#20025]
### Checks

- [X] I have checked that this issue has not already been reported.
- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Reproducible example

```python
import polars as pl
pl.DataFrame({"a": ["01000000000000000000", "02000000000000000000"]}).filter(pl.col("a").str.starts_with("01"))
# shape: (0, 1)
# ┌─────┐
# │ a   │
# │ --- │
# │ str │
# ╞═════╡
# └─────
```

When searching in a shorter string it works:
```python
import polars as pl
pl.DataFrame({"a": ["010000", "0200000"]}).filter(pl.col("a").str.starts_with("01"))
# shape: (1, 1)
# ┌────────┐
# │ a      │
# │ ---    │
# │ str    │
# ╞════════╡
# │ 010000 │
# └────────┘
```

### Issue description

Searching small prefixes in longer strings doesn't work as expected. Probably this relates to recent changes:  https://github.com/pola-rs/polars/pull/19904

Searching small prefixes in short strings works fine.

### Expected behavior

It applies the filter correctly.

### Installed versions

<details>

```
--------Version info---------
Polars:              1.15.0
Index type:          UInt32
Platform:            Linux-6.11.8-300.fc41.x86_64-x86_64-with-glibc2.40
Python:              3.12.7 (main, Oct 11 2024, 09:59:21) [GCC 14.2.1 20240912 (Red Hat 14.2.1-3)]
LTS CPU:             False

----Optional dependencies----
adbc_driver_manager  <not installed>
altair               <not installed>
boto3                <not installed>
cloudpickle          <not installed>
connectorx           <not installed>
deltalake            0.22.0
fastexcel            <not installed>
fsspec               2024.10.0
gevent               <not installed>
google.auth          2.36.0
great_tables         <not installed>
matplotlib           <not installed>
nest_asyncio         1.6.0
numpy                2.1.3
openpyxl             3.1.5
pandas               2.2.3
pyarrow              18.1.0
pydantic             2.10.2
pyiceberg            <not installed>
sqlalchemy           2.0.36
torch                2.5.1+cu124
xlsx2csv             0.8.4
xlsxwriter           <not installed>
```

</details>

**Comment by coastalwhite** [comment]
This is probably already fixed with #20006.

**Comment by alexander-beedie** [comment]
Yup; compiled the latest `head` to check - seems to be fixed:
```python
pl.DataFrame({
  "a": [
    "01000000000000000000",
    "02000000000000000000",
]}).filter(
  pl.col("a").str.starts_with("01"),
)
# shape: (1, 1)
# ┌──────────────────────┐
# │ a                    │
# │ ---                  │
# │ str                  │
# ╞══════════════════════╡
# │ 01000000000000000000 │
# └──────────────────────┘
```
